drivers: i2c: stm32 driver V2 new timing calculation

FRASTM opened the pull request:
This PR is implementing a new formula to calculate the I2C timing value from the I2C clock source and the bit rate.
Algorithm coming from https://github.com/ARMmbed/mbed-os/blob/master/targets/TARGET_STM/i2c_api.c

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/55535

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: i2c: stm32 driver V2 new timing calculation

This PR is implementing a new formula to calculate the I2C timing
value from the I2C clock source and the bit rate.
This is done under flag CONFIG_I2C_STM32_V2_TIMING.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
stm32 I2C V2 FAST bitrate 
![stm32 I2C V2 fast](https://github.com/zephyrproject-rtos/zephyr/assets/51417392/0d46a783-5e1f-4173-9497-68003a6a2d25)

stm32 I2C V2 STANDARD bitrate
![stm32 I2C V2 standard](https://github.com/zephyrproject-rtos/zephyr/assets/51417392/7bc87b8f-b6f0-4991-8021-a60a1d458d97)

FRASTM commented:
Fixing compilation errors

FRASTM commented:
a second commit, to fix the compliance check warning and retrieve the original (correct) function

FRASTM committed:
drivers: i2c: stm32 I2C v2 driver fix compliance with a Macro

Define Macro to fix the compliance check warning :
"DEEP_INDENTATION: Too many leading tabs - consider code refactoring
in the i2c_compute_scll_sclh() function
in the i2c_compute_presc_scldel_sdadel() function

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Rebase, fix macro to get the same timing value
STANDARD : 0xb0c03e40
FAST : 0xc041090f

erwango reviewed (changes requested):
Please cleanup second commit message.
Otherwsie LGTM

FRASTM commented:
> Please cleanup second commit message. Otherwsie LGTM

done

erwango reviewed (changes requested):
Before going further on this point.
Can we clarify why use of [`timings`](https://github.com/zephyrproject-rtos/zephyr/blame/8951bab7e741f3750532d4af68497b9af06a453c/dts/bindings/i2c/st%2Cstm32-i2c-v2.yaml#L23) property didn't help to solve #55535 ?
I'd like first to confirm using this property is not sufficient

erwango reviewed (changes requested):
Since it consumes too much resources, timing computation algo cannot be included in "production" version of the driver.
See https://github.com/zephyrproject-rtos/zephyr/issues/55535#issuecomment-1733180182 for alternatives

FRASTM committed:
drivers: i2c: stm32 for V2 driver get_config with timings

Change the get_config API for the stm32 I2C V2 driver.
It will also return the value of the content of TIMING register
for the I2C V2 bus.
This is hold by a i2c_config_timing structure of the device data

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Add a sample to calculate the I2C V2 timing value and then report to the I2C node in the Device Tree

The i2C V2 driver do not embed the timing calculation code. 
Only when the CONFIG_I2C_STM32_V2_TIMING=y is set by the samples/boards/stm32/i2c_V2_timing/  is the timing value output on the console.

Rebase on 68365d5b67057c8e960a5168c2cf99f4b8c457c1

erwango commented on drivers/i2c/i2c_ll_stm32.c:
```diff
@@ -43,26 +43,60 @@ LOG_MODULE_REGISTER(i2c_ll_stm32);
+{
+	const struct i2c_stm32_config *cfg = dev->config;
+	struct i2c_stm32_data *data = dev->data;
+	struct i2c_config_timing *cfg_timings = (void *)config;
```
I don't think you're supposed to return an other structure than `i2c_stm32_config`.
See i2c.h:
```
 * @param dev_config Pointer to return bit-packed 32-bit value of
 * the I2C controller configuration.
```
It would be logical that the structure to return is the same than the one provided in `i2c_configure`.
Aim is to allow application to compare current and expected configuration:
```
 * i2c_get_config can return cached config or probe hardware, but it has to be
 * up to date with current configuration.
```
This is not a big issue as timing will also be part of this structure

FRASTM replied:
When the the i2C node Device Tree has no <timings> property (i2c_stm32_get_config has noi2c_config_timing and  n_timings == 0) 
Should we return  "(I2C_MODE_CONTROLLER | speed)" but in that case there is no simple way to retrieve the I2C V2 <timing> 32bit value

erwango commented on drivers/i2c/i2c_ll_stm32.c:
```diff
@@ -43,26 +43,60 @@ LOG_MODULE_REGISTER(i2c_ll_stm32);
+	I2C_TypeDef *i2c = cfg->i2c;
+	uint32_t i2c_clock = 0U;
+
+	if (IS_ENABLED(STM32_I2C_DOMAIN_CLOCK_SUPPORT) && (cfg->pclk_len > 1)) {
+		if (clock_control_get_rate(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE),
+					   (clock_control_subsys_t)&cfg->pclken[1],
+					   &i2c_clock) < 0) {
+			LOG_ERR("Failed call clock_control_get_rate(pclken[1])");
+			return -EIO;
+		}
+	} else {
+		if (clock_control_get_rate(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE),
+					   (clock_control_subsys_t) &cfg->pclken[0],
+					   &i2c_clock) < 0) {
+			LOG_ERR("Failed call clock_control_get_rate(pclken[0])");
+			return -EIO;
+		}
+	}
```
Isn't i2c_clock already available in dev->config ?
If not , can it be ?

FRASTM replied:
That was the question : 
i2c_config_timing  is included in the struct i2c_stm32_config but as a constant that cannot be set during the `i2c_stm32_runtime_configure()`

erwango replied:
Ok, good point, I get it now. I need to think about it indeed.

erwango commented on drivers/i2c/i2c_ll_stm32.h:
```diff
@@ -97,4 +98,11 @@ int i2c_stm32_target_register(const struct device *dev, struct i2c_target_config
 int i2c_stm32_target_unregister(const struct device *dev, struct i2c_target_config *config);
 #endif
 
+/* Get the I2C timing register to be reported by the i2c_stm32_get_config */
+#ifdef CONFIG_I2C_STM32_V1
+#define LL_I2C_GET_TIMING(instance) LL_I2C_ReadReg(instance, CCR)
+#else
+#define LL_I2C_GET_TIMING(instance) LL_I2C_ReadReg(instance, TIMINGR)
+#endif
```
To remain consistent with current driver implementation, it would be required to use functions

FRASTM replied:
do you mean 
```
#ifdef CONFIG_I2C_STM32_V1
	cfg_timings->timing_setting = LL_I2C_ReadReg(instance, CCR)
#else
	cfg_timings->timing_setting = LL_I2C_ReadReg(instance, TIMINGR)
#endif

```

erwango replied:
I mean proposing a function `stm32_i2c_get_timing` which will be implemented in `drivers/i2c/i2c_ll_stm32_v1.c` and `drivers/i2c/i2c_ll_stm32_v2.c` files

FRASTM replied:
Version with `struct i2c_config_timing current_timing`  to  store the current timing structure set by runtime config

FRASTM replied:
Get_config : refer to https://github.com/zephyrproject-rtos/zephyr/pull/64422

erwango reviewed (changes requested):
Only reviewed the driver part for now. Some polishing required but globally looks allright

FRASTM committed:
samples: boards: stm32 I2C config and gives V2 timing calculation

This sample reports the I2C configuration of the selected I2C node
from the board DTS : bitrate property of the I2C node
In case of I2C V2, it also gives the TIMINGS which has to be written
in the DTS <timings> property of the I2C V2 node.
Uses the i2c_get_config function with CONFIG_I2C_STM32_V2_TIMING.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Includes the https://github.com/zephyrproject-rtos/zephyr/pull/64422 
Change the sample

FRASTM commented:
rebase on 482416db01e5b8d3e8b91c52bb9daa6eb3d8cddd
after merge of https://github.com/zephyrproject-rtos/zephyr/pull/64422

FRASTM commented:
Running the samples/boards/stm32/i2c_config/ on the b_u585i_iot02a : 

```
*** Booting Zephyr OS build zephyr-v3.5.0-785-g6d149c663d59 ***
I2C timing value, report to the DTS :
timings = <160000000 I2C_BITRATE_FAST  0xC041090F>;
I2C config: I2C_BITRATE_FAST
```

JiiZR commented:
I'am curious what happened to this merge request ? This will fix error https://github.com/zephyrproject-rtos/zephyr/issues/55535.

teburd commented:
@erwango should have been assinged

erwango commented on samples/boards/stm32/i2c_config/README.rst:
```diff
@@ -0,0 +1,78 @@
+    / {
+        aliases {
+                    i2c-0 = &i2c2;
+	 };
```
Please fix indentation.

FRASTM replied:
done

erwango commented on samples/boards/stm32/i2c_config/README.rst:
```diff
@@ -0,0 +1,78 @@
+The calculation of that TIMING value can be given by the `STM32CubeMX`_ application
+**or** by the present sample.
+
+Because the code sequence to calculate the I2C V2 TIMING value is heavy,
+the i2c_ll_stm32_v2.c driver should not embed it.
+That piece of code is activated by the CONFIG_I2C_STM32_V2_TIMING flag
+which must **not be set** except by this sample.
```
```suggestion
Because the code sequence to calculate the I2C V2 TIMING value is heavy,
it is not advised to enable it in production binary.
By enabling CONFIG_I2C_STM32_V2_TIMING flag, this sample allows to 
retrieve timing register configuration for the defined ``clock-frequency``.
User can then configure timing in his application by adapting the following dts
snippet with the output of this sample:
```

FRASTM replied:
done

erwango commented on samples/boards/stm32/i2c_config/README.rst:
```diff
@@ -0,0 +1,78 @@
+.. _i2c_config:
+
+STM32 i2c_config
+################
```
What about making this sample specific to timing V2 only ?
```suggestion
STM32 I2C V2 timings
####################
```

For V1, this sample doesn't add much value and this would simplify the sample configuration a bit.

FRASTM replied:
done

FRASTM commented:
rebase on 8100871856ff7394f9004457ade321d7fc3bf240
Renaming the samples"I2C timing"
Availability is I2C V2 only,  setting CONFIG_I2C_STM32_V2_TIMING=y in the prj.conf

erwango commented:
@gautierg-st PTAL

marwaiehm-st commented on samples/boards/stm32/i2c_timing/src/main.c:
```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2023 STMicroelectronics
+ *
```
Copyright (c) 2024 STMicroelectronics

FRASTM replied:
this PR started last year, so 2023 was pertinent at that time,
should we really  update this comment since this is a "internal" sample

marwaiehm-st reviewed (approved):
LGTM

teburd commented:
Needs CI fixes

mathieuchopstm commented on drivers/i2c/i2c_ll_stm32.c:
```diff
@@ -56,6 +56,23 @@ int i2c_stm32_get_config(const struct device *dev, uint32_t *config)
+		printk(" I2C_SPEED_FAST_PLUS ");
+	}
+
+	printk(" 0x%X>;\n", data->current_timing.timing_setting);
```
Maybe these `printk`s should be changed to `LOG_xxx`s instead?

ajarmouni-st reviewed (changes requested):
change `printk` to `LOG_XXX` in driver code

FRASTM commented:
rebase on 2efc447c1b41efd23447ea180731d673114f06d8
Version with **LOG_INF** instead of printk in the driver
 --> change the  sample.yaml
--> add the LOG_INF to the prj.conf
```
*** Booting Zephyr OS build v3.6.0-5746-ge94cfcf43a1c ***
[00:00:09.192,000] <inf> i2c_ll_stm32: I2C timing value, report to the DTS :
[00:00:09.192,000] <inf> i2c_ll_stm32: timings = <32000000 I2C_BITRATE_FAST 0xB01626>;
I2C freq. I2C_BITRATE_FAST
```
